File: scripts/python/src/tools/utils/sysinfo/bench/record.py
import hashlib
import statistics
from dataclasses import dataclass, field, replace

from tools.utils.sysinfo.formatting import as_dict, as_list

SCHEMA = 1

HIB = "HIB"
LIB = "LIB"

WORLD = "world"
PLATFORM = "platform"
HOST = "host"

CLEAN = "clean"
NOISY = "noisy"
ABORTED = "aborted"

TIERS = ("quick", "standard", "heavy")


@dataclass(frozen=True)
class Metric:
    key: str
    method: str
    scale: str
    proportion: str
    comparable: str
    tool: str = ""
    tool_version: str = ""
    samples: tuple[float, ...] = ()
    detail: dict = field(default_factory=dict)

    @property
    def times_to_run(self):
        return len(self.samples)

    @property
    def median(self):
        if not self.samples:
            return None
        return statistics.median(self.samples)

    @property
    def mad(self):
        if not self.samples:
            return None
        center = statistics.median(self.samples)
        return statistics.median([abs(value - center) for value in self.samples])

    @property
    def rsd_pct(self):
        return relative_deviation(self.samples)

    @property
    def family(self):
        return self.key.split(".", 1)[0]

    def better(self, other):
        if self.median is None or other is None:
            return None
        if self.proportion == LIB:
            return self.median < other
        return self.median > other

    def to_json(self):
        payload = {
            "key": self.key,
            "method": self.method,
            "tool": self.tool,
            "tool_version": self.tool_version,
            "scale": self.scale,
            "proportion": self.proportion,
            "comparable": self.comparable,
            "times_to_run": self.times_to_run,
            "samples": list(self.samples),
            "median": self.median,
            "mad": self.mad,
            "rsd_pct": self.rsd_pct,
        }
        if self.detail:
            payload["detail"] = self.detail
        return payload

    @staticmethod
    def from_json(payload):
        return Metric(
            key=payload.get("key", ""),
            method=payload.get("method", ""),
            scale=payload.get("scale", ""),
            proportion=payload.get("proportion", HIB),
            comparable=payload.get("comparable", HOST),
            tool=payload.get("tool", ""),
            tool_version=payload.get("tool_version", ""),
            samples=tuple(payload.get("samples") or ()),
            detail=as_dict(payload.get("detail")),
        )


@dataclass(frozen=True)
class Run:
    run_id: str
    host: str
    started: str
    tier: str
    grade: str
    snapshot: dict = field(default_factory=dict)
    install: dict = field(default_factory=dict)
    conditions: dict = field(default_factory=dict)
    metrics: tuple[Metric, ...] = ()
    note: str = ""
    tags: tuple[str, ...] = ()
    dotfiles_sha: str = ""
    gate_reasons: tuple[str, ...] = ()
    bytes_written: int = 0
    schema: int = SCHEMA

    @property
    def epoch(self):
        return epoch_of(self.snapshot)

    @property
    def os_id(self):
        return self.install.get("os", "")

    def metric(self, key):
        for entry in self.metrics:
            if entry.key == key:
                return entry
        return None

    def to_json(self):
        return {
            "schema": self.schema,
            "run_id": self.run_id,
            "host": self.host,
            "epoch": self.epoch,
            "started": self.started,
            "tier": self.tier,
            "grade": self.grade,
            "note": self.note,
            "tags": list(self.tags),
            "dotfiles_sha": self.dotfiles_sha,
            "gate_reasons": list(self.gate_reasons),
            "bytes_written": self.bytes_written,
            "snapshot": self.snapshot,
            "install": self.install,
            "conditions": self.conditions,
            "metrics": [metric.to_json() for metric in self.metrics],
        }

    @staticmethod
    def from_json(payload):
        return Run(
            run_id=payload.get("run_id", ""),
            host=payload.get("host", ""),
            started=payload.get("started", ""),
            tier=payload.get("tier", ""),
            grade=payload.get("grade", ""),
            snapshot=as_dict(payload.get("snapshot")),
            install=as_dict(payload.get("install")),
            conditions=as_dict(payload.get("conditions")),
            metrics=tuple(Metric.from_json(item) for item in as_list(payload.get("metrics"))),
            note=payload.get("note", ""),
            tags=tuple(payload.get("tags") or ()),
            dotfiles_sha=payload.get("dotfiles_sha", ""),
            gate_reasons=tuple(payload.get("gate_reasons") or ()),
            bytes_written=payload.get("bytes_written") or 0,
            schema=payload.get("schema") or SCHEMA,
        )


def relative_deviation(samples):
    values = [value for value in samples if isinstance(value, (int, float))]
    if len(values) < 2:
        return 0.0
    mean = statistics.fmean(values)
    if not mean:
        return 0.0
    return abs(statistics.stdev(values) / mean) * 100


def method_series(method):
    name, _, version = method.partition("/")
    parts = version.split(".")
    major = parts[0] if parts else ""
    minor = parts[1] if len(parts) > 1 else ""
    return f"{name}/{major}.{minor}"


def comparable_methods(left, right):
    return method_series(left.method) == method_series(right.method)


GIB = 1024**3


def whole_gib(value):
    """Capacity rounded down to GiB, as a string.

    Capacities arrive from several sources at several precisions -- VRAM comes
    from nvidia-smi as a float and from fastfetch as an int, and the two differ
    by a few hundred MiB for the same card. Comparing the raw values meant a
    three second nvidia-smi timeout silently changed the machine's identity and
    orphaned its pinned baseline, after which real regressions went unreported.
    """
    try:
        return str(int(float(value) // GIB))
    except (TypeError, ValueError):
        return ""


def identity_fields(snapshot):
    cpu = as_dict(snapshot.get("cpu"))
    memory = as_dict(snapshot.get("memory"))
    values = [
        cpu.get("model", ""),
        str(cpu.get("cores_physical") or ""),
        str(cpu.get("cores_logical") or ""),
        whole_gib(memory.get("total")),
    ]
    # Sorted, because device enumeration order is not identity: the same two
    # drives reported in the other order used to read as a different machine.
    # memory.modules is deliberately absent -- it reads 0 without root and 2
    # with, so one sudo run was enough to orphan the baseline.
    values.extend(
        sorted(
            f"{gpu.get('name', '')}:{whole_gib(gpu.get('memory_total'))}"
            for gpu in as_list(snapshot.get("gpu"))
        )
    )
    values.extend(
        sorted(
            f"{disk.get('name', '')}:{whole_gib(disk.get('size'))}"
            for disk in as_list(snapshot.get("disks"))
        )
    )
    return values


def epoch_of(snapshot):
    joined = "\n".join(identity_fields(snapshot))
    return hashlib.blake2s(joined.encode("utf-8"), digest_size=4).hexdigest()


def snapshot_differences(left, right):
    changes = []
    for label, key in (("CPU", "cpu"), ("Memory", "memory")):
        before, after = as_dict(left.get(key)), as_dict(right.get(key))
        for name in ("model", "total", "modules", "cores_physical"):
            if (name in before or name in after) and before.get(name) != after.get(name):
                changes.append((label, name, before.get(name), after.get(name)))
    for label, key, name in (("GPU", "gpu", "name"), ("Storage", "disks", "name")):
        before = [item.get(name, "") for item in as_list(left.get(key))]
        after = [item.get(name, "") for item in as_list(right.get(key))]
        if before != after:
            changes.append((label, name, ", ".join(before), ", ".join(after)))
    return tuple(changes)


def with_metrics(run, metrics):
    return replace(run, metrics=tuple(metrics))

File: scripts/rust/crates/sysinfo/src/bench/record.rs
use blake2::digest::{Update, VariableOutput};
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

pub const HIB: &str = "HIB";
pub const LIB: &str = "LIB";
pub const WORLD: &str = "world";
pub const HOST: &str = "host";
pub const CLEAN: &[&str] = &["clean"];
pub const ANY: &[&str] = &["clean", "noisy", "aborted"];
pub const TIERS: [&str; 3] = ["quick", "standard", "heavy"];

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(default)]
pub struct Metric {
    pub key: String,
    pub method: String,
    pub scale: String,
    pub proportion: String,
    pub comparable: String,
    pub tool: String,
    pub tool_version: String,
    pub samples: Vec<f64>,
    pub detail: Value,
}

impl Default for Metric {
    fn default() -> Self {
        Self {
            key: String::new(),
            method: String::new(),
            scale: String::new(),
            proportion: HIB.into(),
            comparable: HOST.into(),
            tool: String::new(),
            tool_version: String::new(),
            samples: Vec::new(),
            detail: json!({}),
        }
    }
}

impl Metric {
    pub fn median(&self) -> Option<f64> {
        median(&self.samples)
    }
    pub fn mad(&self) -> Option<f64> {
        let center = self.median()?;
        median(
            &self
                .samples
                .iter()
                .map(|value| (value - center).abs())
                .collect::<Vec<_>>(),
        )
    }
    pub fn rsd_pct(&self) -> f64 {
        relative_deviation(&self.samples)
    }
    pub fn family(&self) -> &str {
        self.key.split('.').next().unwrap_or("")
    }
    pub fn to_json(&self) -> Value {
        let mut value = json!({
            "key": self.key,
            "method": self.method,
            "tool": self.tool,
            "tool_version": self.tool_version,
            "scale": self.scale,
            "proportion": self.proportion,
            "comparable": self.comparable,
            "times_to_run": self.samples.len(),
            "samples": self.samples,
            "median": self.median(),
            "mad": self.mad(),
            "rsd_pct": self.rsd_pct(),
        });
        if self
            .detail
            .as_object()
            .is_some_and(|detail| !detail.is_empty())
        {
            value["detail"] = self.detail.clone();
        }
        value
    }
}

impl Serialize for Metric {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        self.to_json().serialize(serializer)
    }
}

#[derive(Clone, Debug, Deserialize, PartialEq)]
#[serde(default)]
pub struct Run {
    pub run_id: String,
    pub host: String,
    pub started: String,
    pub tier: String,
    pub grade: String,
    pub snapshot: Value,
    pub install: Value,
    pub conditions: Value,
    pub metrics: Vec<Metric>,
    pub note: String,
    pub tags: Vec<String>,
    pub dotfiles_sha: String,
    pub gate_reasons: Vec<String>,
    pub bytes_written: u64,
    pub schema: u32,
}

impl Default for Run {
    fn default() -> Self {
        Self {
            run_id: String::new(),
            host: String::new(),
            started: String::new(),
            tier: String::new(),
            grade: String::new(),
            snapshot: json!({}),
            install: json!({}),
            conditions: json!({}),
            metrics: Vec::new(),
            note: String::new(),
            tags: Vec::new(),
            dotfiles_sha: String::new(),
            gate_reasons: Vec::new(),
            bytes_written: 0,
            schema: 1,
        }
    }
}

impl Run {
    pub fn epoch(&self) -> String {
        epoch_of(&self.snapshot)
    }
    pub fn metric(&self, key: &str) -> Option<&Metric> {
        self.metrics.iter().find(|metric| metric.key == key)
    }
    pub fn os_id(&self) -> &str {
        self.install["os"].as_str().unwrap_or("")
    }
    pub fn to_json(&self) -> Value {
        json!({
            "schema": self.schema,
            "run_id": self.run_id,
            "host": self.host,
            "epoch": self.epoch(),
            "started": self.started,
            "tier": self.tier,
            "grade": self.grade,
            "note": self.note,
            "tags": self.tags,
            "dotfiles_sha": self.dotfiles_sha,
            "gate_reasons": self.gate_reasons,
            "bytes_written": self.bytes_written,
            "snapshot": self.snapshot,
            "install": self.install,
            "conditions": self.conditions,
            "metrics": self.metrics,
        })
    }
}
impl Serialize for Run {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        self.to_json().serialize(serializer)
    }
}

pub fn median(samples: &[f64]) -> Option<f64> {
    if samples.is_empty() {
        return None;
    }
    let mut values = samples.to_vec();
    values.sort_by(f64::total_cmp);
    let middle = values.len() / 2;
    Some(if values.len().is_multiple_of(2) {
        (values[middle - 1] + values[middle]) / 2.0
    } else {
        values[middle]
    })
}
pub fn relative_deviation(samples: &[f64]) -> f64 {
    if samples.len() < 2 {
        return 0.0;
    }
    let mean = samples.iter().sum::<f64>() / samples.len() as f64;
    if mean == 0.0 {
        return 0.0;
    }
    let variance = samples
        .iter()
        .map(|value| (value - mean).powi(2))
        .sum::<f64>()
        / (samples.len() - 1) as f64;
    (variance.sqrt() / mean).abs() * 100.0
}
pub fn method_series(method: &str) -> String {
    let (name, version) = method.split_once('/').unwrap_or((method, ""));
    let mut parts = version.split('.');
    format!(
        "{name}/{}.{}",
        parts.next().unwrap_or(""),
        parts.next().unwrap_or("")
    )
}
pub fn whole_gib(value: &Value) -> String {
    value
        .as_f64()
        .or_else(|| value.as_str()?.parse().ok())
        .filter(|n| n.is_finite())
        .map(|n| format!("{:.0}", (n / 1073741824.0).floor()))
        .unwrap_or_default()
}
fn text(value: &Value) -> String {
    value.as_str().map(str::to_owned).unwrap_or_else(|| {
        if value.is_null() || value.as_i64() == Some(0) {
            String::new()
        } else {
            value.to_string()
        }
    })
}
pub fn identity_fields(snapshot: &Value) -> Vec<String> {
    let cpu = &snapshot["cpu"];
    let mut values = vec![
        text(&cpu["model"]),
        text(&cpu["cores_physical"]),
        text(&cpu["cores_logical"]),
        whole_gib(&snapshot["memory"]["total"]),
    ];
    for (key, size) in [("gpu", "memory_total"), ("disks", "size")] {
        let mut devices = snapshot[key]
            .as_array()
            .into_iter()
            .flatten()
            .map(|device| format!("{}:{}", text(&device["name"]), whole_gib(&device[size])))
            .collect::<Vec<_>>();
        devices.sort();
        values.extend(devices);
    }
    values
}
pub fn epoch_of(snapshot: &Value) -> String {
    let mut hash = blake2::Blake2sVar::new(4).expect("valid digest length");
    hash.update(identity_fields(snapshot).join("\n").as_bytes());
    let mut bytes = [0; 4];
    hash.finalize_variable(&mut bytes)
        .expect("fixed digest buffer");
    bytes.iter().map(|byte| format!("{byte:02x}")).collect()
}
pub type Change = (String, String, Value, Value);
pub fn snapshot_differences(left: &Value, right: &Value) -> Vec<Change> {
    let mut changes = Vec::new();
    for (label, key) in [("CPU", "cpu"), ("Memory", "memory")] {
        for name in ["model", "total", "modules", "cores_physical"] {
            if left[key][name] != right[key][name] {
                changes.push((
                    label.into(),
                    name.into(),
                    left[key][name].clone(),
                    right[key][name].clone(),
                ));
            }
        }
    }
    for (label, key) in [("GPU", "gpu"), ("Storage", "disks")] {
        let names = |value: &Value| {
            value[key]
                .as_array()
                .into_iter()
                .flatten()
                .map(|item| text(&item["name"]))
                .collect::<Vec<_>>()
        };
        let before = names(left);
        let after = names(right);
        if before != after {
            changes.push((
                label.into(),
                "name".into(),
                json!(before.join(", ")),
                json!(after.join(", ")),
            ));
        }
    }
    changes
}
